### ISSUE TITLE
[Bug] Mixed tensor and scalar stores to the same GM tensor can silently corrupt a cache line

### ISSUE THREAD

**Issue by high-cloud** [hw-native-sys/pypto#2005]
## Background

In pypto-lib, the DeepSeek-V4 prefill compressor on a2a3 silently corrupts
`write_dst_map` when whole-tensor initialization and scalar overrides target
the same GM tensor in one `pl.at` scope.

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `85f86efe` |
| pypto | `12aa8bdb` (branch: main) |
| simpler | `c94aa9f3` (detached) |
| ptoas | `0.48` |
| pto-isa | `83d01313` |
| CANN | `9.0.0` |
| Python | `3.10.20` |

All downstream versions match the pins from the PyPTO checkout.

Diagnosis: **PyPTO Codegen**. PyPTO accepts this source pattern and lowers it
to `pto.tstore` plus `pto.store_scalar` on overlapping GM cache lines,
without a coherence/order guarantee or diagnostic.

### Component

Codegen

### Description

A tensor slice assignment followed by `pl.write` to the same GM tensor can
silently produce incorrect contents:

```python
with pl.at(level=pl.Level.CORE_GROUP):
    output[0:1, 0:32] = pl.full([1, 32], dtype=pl.INT32, value=-1)
    for i in pl.range(4):
        pl.write(output, [0, i], pl.cast(i, pl.INT32))
```

The generated AIV kernel contains:

```cpp
TSTORE(dst, full_minus_one_tile);  // PIPE_MTE3
// No MTE3-to-scalar ordering/coherence operation here.
output[i] = i;                     // scalar/D-cache store
pipe_barrier(PIPE_ALL);
dcci((__gm__ void*)0, ENTIRE_DATA_CACHE, CACHELINE_OUT);
```

The trailing barrier waits for both pipelines but does not establish their
relative write order or prevent D-cache writeback from overwriting bytes
written by MTE3.

### Steps to Reproduce

Save the following as `mixed_tstore_scalar_repro.py` in a pypto-lib checkout:

<details>
<summary>Minimal reproduction script</summary>

```python
import argparse
import pypto.language as pl
from golden import TensorSpec, run_jit

N = 32
VALID = 4
LAYERS = 20
SCAN = 128

@pl.jit
def mixed_tstore_scalar(
    output: pl.Out[pl.Tensor[[LAYERS, N], pl.INT32]],
):
    scratch = pl.create_tensor([1, N], dtype=pl.INT32)
    for layer in pl.range(LAYERS):
        with pl.at(level=pl.Level.CORE_GROUP, name_hint="mixed_tstore_scalar"):
            scratch[0:1, 0:N] = pl.full([1, N], dtype=pl.INT32, value=-1)
            seen = pl.cast(0, pl.INDEX)
            for scan_i in pl.range(SCAN):
                if scan_i < 16:
                    if (scan_i + 1) % 4 == 0:
                        pl.write(scratch, [0, seen], pl.cast(seen, pl.INT32))
                        seen = seen + 1
        with pl.at(level=pl.Level.CORE_GROUP, name_hint="observe_mixed_store"):
            output[layer : layer + 1, 0:N] = scratch[0:1, 0:N]
    return output

def build_specs():
    import torch
    return [TensorSpec(
        "output", [LAYERS, N], torch.int32,
        init_value=lambda: torch.zeros([LAYERS, N], dtype=torch.int32),
        is_output=True,
    )]

def golden(values):
    import torch
    values["output"][:] = -1
    values["output"][:, 0:VALID] = torch.arange(VALID, dtype=torch.int32)

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", default="a2a3")
    parser.add_argument("-d", "--device", type=int, default=0)
    args = parser.parse_args()
    result = run_jit(
        fn=mixed_tstore_scalar,
        specs=build_specs(),
        golden_fn=golden,
        runtime_cfg={"platform": args.platform, "device_id": args.device},
        rtol=0,
        atol=0,
    )
    if not result.passed:
        raise SystemExit(1)
```
</details>

Run:

```bash
task-submit --ptoas 0.48 --device auto --max-time 120 --timeout 180 \
  --run "python mixed_tstore_scalar_repro.py -p a2a3 -d {}"
```

The minimal repro failed 3 out of 4 runs. Failures vary by timing:

```text
# TSTORE wins after scalar writes: scalar overrides are lost.
actual=-1, expected=0
actual=-1, expected=1
actual=-1, expected=2
actual=-1, expected=3

# Scalar D-cache writeback wins: untouched tail bytes revert to stale zeros.
indices 4..15: actual=0, expected=-1
```

The generated PTO IR contains one `pto.tstore` and one
`pto.store_scalar` site targeting the same pointer.

The original DeepSeek-V4 EP2, `num_tokens=16` tensor-dump repro observed:

```text
expected: [0, 1, 2, 3, -1, -1, ..., -1]
actual:   [0, 1, 2, 3,  0,  0, ...,  0, -1, ..., -1]
                            indices 4..15
```

Rank 0 had 400 bad consumer snapshots and rank 1 had 132. The corruption
persisted through pool, RMS/rope, and cache-write consumers.

### Expected Behavior

PyPTO should preserve source-order memory semantics when tensor and scalar
stores overlap. It should either:

1. lower both writes through a coherent store path;
2. insert the required MTE3/scalar and D-cache synchronization; or
3. reject the unsupported pattern or emit a clear compile-time warning.

### Actual Behavior

Compilation succeeds without a warning and intermittently produces corrupted
GM data. Replacing tensor initialization with scalar `pl.write`
initialization eliminates `pto.tstore`; the same model dump check then
reports zero mismatches across 8,232 position-map and 8,400 destination-map
snapshots.

### Git Commit ID

`12aa8bdb2de1a8e519c2d01d32a8bd9d05cb0b42`

### NPU Kind

Ascend 910B

### Host Platform

Linux (aarch64)

### Additional Context

The same unsafe lowering occurs in both `prefill_c4_write_map` and
`prefill_idx_c4_write_map`.

Related: #1561, which covers remote `TSTORE -> TNOTIFY` ordering rather than
local GM `TSTORE -> store_scalar` cache coherence.

**Comment by TaoTao-real** [comment]
补充一个 A3 最小实验结果，验证 “MTE3 整行写 GM 后，再通过 scalar 局部写同一条 64B GM line，是否可以靠 `dcci/dsb` 组合修正” 这个方向。

实验结构：

- 数据按 256 条 64B line 组织，每条 line 是 16 个 `int32`。
- 初始 GM 输出全 0。
- `TSTORE` 通过 MTE3 把整条 64B line 写成 `-1`。
- 随后 scalar 只写前 4 个 `int32` 为 `0, 1, 2, 3`。
- 期望每条 line 最终为：`[0, 1, 2, 3, -1, -1, ...]`。
- 在每条 line 开始前先 scalar read `line[15]`，用于模拟 scalar 侧已经持有旧 line 状态。
- 通过 A3 `task-submit` 跑多轮验证。

测试过的序列和结果：

1. 无前置维护的控制组：

```cpp
TSTORE(...);
pipe_barrier(PIPE_MTE3);
dsb(DSB_DDR);
scalar store line[0..3];
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果：5/5 失败，`bad_elems` 约 3064 到 3072。典型结果是前 4 个 scalar 写成功，但后 12 个元素被旧的 0 覆盖：

```text
[0, 1, 2, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```

2. MTE3 后、scalar 写前增加一次 `dcci + dsb`：

```cpp
TSTORE(...);
pipe_barrier(PIPE_MTE3);
dsb(DSB_DDR);
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
scalar store line[0..3];
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果：仍然失败。错误大幅减少，但不是 0，常见为 `bad_elems=4/12/16`，集中在 line 64/65 附近。

3. 按 “没有纯 dci 时，MTE3 前后都做 dcci” 的思路测试：

```cpp
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
TSTORE(...);
pipe_barrier(PIPE_MTE3);
dsb(DSB_DDR);
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
scalar store line[0..3];
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果：5/5 失败，`bad_elems=32`，错误稳定出现在 0、64、128、192 附近。例如 line 0 变成整条 `-1`，说明 scalar 局部写没有可靠落到最终 GM：

```text
actual:   [-1, -1, -1, -1, -1, ...]
expected: [0, 1, 2, 3, -1, ...]
```

4. 只在 MTE3 前做一次 `dcci + dsb`：

```cpp
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
TSTORE(...);
pipe_barrier(PIPE_MTE3);
dsb(DSB_DDR);
scalar store line[0..3];
dcci(line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果：5/5 失败，`bad_elems` 约 104 到 248，比 MTE3 后加 dcci 更差。

当前结论：

- 这个问题不是单纯的 pipe ordering 问题；`pipe_barrier(PIPE_MTE3)` 和 `dsb(DSB_DDR)` 只能保证 MTE3 写完成和可见，不能保证后续 scalar partial store 会以 MTE3 写出的 64B line 作为 merge base。
- 仅靠现有 `dcci` 组合没有测出可靠正确的序列。`dcci` 如果包含 clean/writeback 语义，那么它不是纯 invalidate，可能把 scalar 侧旧 line 写回去；而 scalar partial store 又没有 byte-level mask/coherent merge 保证。
- 因此目前不能把 “MTE3 整行写 + scalar 局部写同一条 64B GM cache line” 作为合法可自动修复的模式。

建议：

- PyPTO 侧优先避免 TSTORE/MTE3 和 scalar store 混写同一条 64B GM cache line。
- scalar metadata 或局部更新字段建议独立 64B 对齐隔离，避免和 tensor payload 共 line。
- 如果业务确实需要混写，应改成同一路径负责整条 64B line，或者明确依赖硬件提供纯 invalidate `dci`/byte-mask merge 等能力后再重新定义合法序列。
- PTOAS 的 memory consistency pass 可以保证 signal/payload 的 release/acquire 和 pipe drain，但无法仅通过 fence/dcci 自动修复这个 same-line mixed writer 问题。

**Comment by TaoTao-real** [comment]
补充/更正后续 A3 实验结果：前面提到的 `MTE3 -> PIPE_S` illegal 是测试用例不干净导致的误判。修正最小 pipe-pair 测试后，`set_flag(PIPE_MTE3, PIPE_S, EVENT_ID0..7); wait_flag(PIPE_MTE3, PIPE_S, EVENT_ID0..7);` 在 A3 上可以正常执行。

最小复现场景仍然是：

```text
TSTORE/MTE3 整条 64B line 写 GM 为 -1
随后 scalar store 只写同一条 line 的前 4 个 int32 为 0,1,2,3
期望最终 line = [0, 1, 2, 3, -1, -1, ...]
```

关键实验结果如下。

1. `pipe_barrier(PIPE_MTE3)` 方向不闭合。

```cpp
TSTORE_IMPL(dst, tile);
pipe_barrier(PIPE_MTE3);
dsb(DSB_DDR);
scalar store line[0..3];
dcci((__gm__ void *)line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果失败，典型 `bad_elems=3072`。前 4 个 scalar 写成功，但后 12 个元素被旧的 0 覆盖：

```text
actual:   [0, 1, 2, 3, 0, 0, ...]
expected: [0, 1, 2, 3, -1, -1, ...]
```

2. `MTE3 -> PIPE_S set/wait` 本身可以执行，但单独不能修复。

```cpp
TSTORE_IMPL(dst, tile);
set_flag(PIPE_MTE3, PIPE_S, EVENT_ID0);
wait_flag(PIPE_MTE3, PIPE_S, EVENT_ID0);
dsb(DSB_DDR);
scalar store line[0..3];
dcci((__gm__ void *)line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果仍失败，`bad_elems=3072`。也就是说，`MTE3 -> S` 能提供跨 pipe 顺序，但不能替代 scalar 侧对同一条 GM line 的 cache/write-buffer 维护。

3. `MTE3 -> PIPE_S set/wait + scalar 写前 dcci/dsb` 可以修复当前最小复现。

```cpp
TSTORE_IMPL(dst, tile);
set_flag(PIPE_MTE3, PIPE_S, EVENT_ID0);
wait_flag(PIPE_MTE3, PIPE_S, EVENT_ID0);
dsb(DSB_DDR);
dcci((__gm__ void *)line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
scalar store line[0..3];
dcci((__gm__ void *)line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果：initial run + 5 次重复运行均通过，`bad_elems=0`。

4. 保守的 `PIPE_ALL + dcci/dsb` 方案也可以修复当前最小复现。

```cpp
TSTORE_IMPL(dst, tile);
pipe_barrier(PIPE_ALL);
dsb(DSB_DDR);
dcci((__gm__ void *)line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
scalar store line[0..3];
dcci((__gm__ void *)line, SINGLE_CACHE_LINE);
dsb(DSB_DDR);
```

结果同样稳定通过，`bad_elems=0`。

5. 诊断性 `MTE3 -> MTE2 set/wait + dcci/dsb` 不可靠，只能把错误数量降低到少量，不能完全闭合。

当前结论更新为：

- 该问题不是普通的 MTE3 pipe 内部 WAW/RAW 顺序问题，`pipe_barrier(PIPE_MTE3)` 不够。
- `MTE3 -> PIPE_S set/wait` 在 A3 上是可执行的，并且可以承担 “scalar 侧等待 MTE3 完成” 这一部分顺序约束。
- 但 #2005 的核心还包括 scalar partial store 对同一 64B GM line 的 cache/write-buffer 状态问题，所以在 scalar 写之前仍需要 `dcci(line) + dsb`，scalar 写之后也需要 `dcci(line) + dsb` 把结果推出去。
- 因此当前最小可行候选序列是：

```text
TSTORE
set/wait(PIPE_MTE3, PIPE_S)
dsb
dcci(line)
dsb
scalar store
dcci(line)
dsb
```

- 如果不想依赖 `MTE3 -> PIPE_S` 这个更窄的 pair，保守序列可以继续使用：

```text
TSTORE
pipe_barrier(PIPE_ALL)
dsb
dcci(line)
dsb
scalar store
dcci(line)
dsb
```

建议 PyPTO 优先避免 MTE3 full-line store 和 scalar partial store 混写同一条 64B line；如果业务确实需要混写，需要生成上述显式维护序列，并建议再和硬件/ISA 同事确认正式推荐的最小序列。

实验目录：

```text
pipe-pair 合法性测试：/home/zhongxuan/ptoas_board_runs/_tmp_mte3_s_pipepair_tassign_20260716_201901
数据正确性测试：/home/zhongxuan/ptoas_board_runs/_tmp_mte_scalar_mte3_s_20260716_202350
```
